Add ability to cancel the progress

boyvanamstel opened the pull request:
Brings over the cancel functionality added to Gifski recently.

![cancel](https://user-images.githubusercontent.com/225410/51617966-a2649180-1f2d-11e9-8049-dd826cc65c86.gif)

I chose to copy over the `CustomButton` verbatim. Using it here might help fleshing out how it should behave as a standalone component later.

The cancel button currently only appears when a `Progress` instance has been set that is `isCancellable`.

boyvanamstel committed:
Add cancel button

boyvanamstel committed:
Update example

boyvanamstel committed:
Tweak hover rect

boyvanamstel committed:
Update readme

sindresorhus commented:
If I click the cancel button, I don't think we should reset the progress. but rather keep the progress. And maybe reduce the opacity of the label and progressbar to indicate it's disabled? Or maybe even better, reduce saturation?

sindresorhus commented:
Unrelated to this PR, but I really wish we could do a better checkmark icon. The current one doesn't fit in at all with the nice and thin `X`.

sindresorhus commented:
I think we should support the cancel button when `.progressValue` is used too. There are many cases where using the `Progress` API doesn't make sense. Then we need `.isCancellable` and `.onCancel` on the CircularProgress instance. If a `Progress` instance is specified, the `CircularProgress#isCancellable` takes priority over `progress.isCancellable`.

sindresorhus commented on Sources/CircularProgress/CustomButton.swift:
```diff
@@ -0,0 +1,327 @@
+
+extension CALayer {
+	// TODO: Find a way to use a strongly-typed KeyPath here.
+	// TODO: Accept NSColor instead of CGColor.
```
If you have any ideas how to solve these, I would appreciate the help. No worries if not though.

sindresorhus committed:
Update readme.md

boyvanamstel commented:
> Unrelated to this PR, but I really wish we could do a better checkmark icon. The current one doesn't fit in at all with the nice and thin `X`.

Yea, I know. It bothers me too. There might not be a way around adding a `successLayer`.

boyvanamstel commented:
> I think we should support the cancel button when `.progressValue` is used too. There are many cases where using the `Progress` API doesn't make sense. Then we need `.isCancellable` and `.onCancel` on the CircularProgress instance. If a `Progress` instance is specified, the `CircularProgress#isCancellable` takes priority over `progress.isCancellable`.

Makes sense. I added `.onCancel` in one of the commits, so I can just bring that back. I kinda like how simple the API is by just leveraging `Progress`, but I trust you when you say there are use cases where `Progress` isn't the best option.

sindresorhus commented:
> Yea, I know. It bothers me too. There might not be a way around adding a successLayer.

Moved the discussion to a separate issue: https://github.com/sindresorhus/CircularProgress/issues/5

boyvanamstel committed:
Re-commit shared schemes (fixes no schemes in AppCode)

boyvanamstel committed:
Implement cancellation for all progress types

boyvanamstel committed:
Update documentation

boyvanamstel committed:
Reduce opacity in Example on cancellation

boyvanamstel commented:
I've added the option to cancel either manual or progress based views.

Instead of `.onCancel` I've added `.onCancelled` because it only fires when cancellation was successful.

Cancellation is disabled by default to not break existing projects.

I've made `.isCancelled` and `.isFinished` `@IBInspectible` so you can see what they look like. Currently `.isCancelled` doesn't change the appearance, I've implemented an opacity change in the example project through `.isCancelled`. We could provide a default state through desaturation or an opacity change later.

![cancellable](https://user-images.githubusercontent.com/225410/51679180-7a813680-1fde-11e9-844e-fe8e4cdad342.gif)

sindresorhus commented on Example/AppDelegate.swift:
```diff
@@ -23,12 +23,13 @@ final class AppDelegate: NSObject, NSApplicationDelegate {
-		manualCircularProgress.onCancelled = {
-			print("Manual cancelled!")
+		manualCircularProgress.onCancelled = { circularProgress in
+			circularProgress.alphaValue = 0.3
```
Why not just `manualCircularProgress.alphaValue = 0.3`? I usually pass on self in closures too, but is it really that beneficial?

boyvanamstel replied:
I was juggling a bunch of circular progresses in the example app at some point. It cleaned up the code a bit then. I think it's better to not have it moving forward.

sindresorhus committed:
Update CircularProgress.swift

sindresorhus committed:
Update util.swift

sindresorhus committed:
Update readme.md

sindresorhus commented on readme.md:
```diff
@@ -119,6 +121,31 @@ Reset the progress back to zero without animating.
+/**
+Triggers when the progress was cancelled succesfully.
+*/
+var onCancelled: (() -> Void)?
```
This needs to be updated if we decide to keep the parameter: https://github.com/sindresorhus/CircularProgress/pull/4#discussion_r250964699

boyvanamstel replied:
Parameter has been removed.

sindresorhus committed:
Update AppDelegate.swift

boyvanamstel committed:
Remove reference to self from onCancelled closure

sindresorhus commented on Sources/CircularProgress/CircularProgress.swift:
```diff
@@ -61,22 +74,67 @@ public final class CircularProgress: NSView {
 				}
+
+				finishedObserver = progressInstance.observe(\.isFinished) { sender, _ in
+					guard !sender.isCancelled && sender.isFinished else {
```
```suggestion
					guard !self.isCancelled && sender.isFinished else {
```

For consistency?

boyvanamstel committed:
Fix consistency in fraction completed observer

boyvanamstel commented:
Rebased on master.

This is a warning now btw. Do you know how to fix it, besides putting it back on a single line?

<img width="1817" alt="screenshot 2019-01-27 at 10 43 40" src="https://user-images.githubusercontent.com/225410/51799329-7fcdc380-2220-11e9-8376-f0c75326e59a.png">

sindresorhus committed:
Update AppDelegate.swift

sindresorhus committed:
Update AppDelegate.swift

sindresorhus commented:
> Do you know how to fix it, besides putting it back on a single line?

https://github.com/sindresorhus/CircularProgress/pull/4/commits/1e9bb8b2b59610bc8ce09d4d7913e75112945599

sindresorhus commented:
Yay! This is awesome. 🎉

sindresorhus commented:
@boyvanamstel Do you want to bring these changes back to Gifski or should I?

boyvanamstel commented:
> @boyvanamstel Do you want to bring these changes back to Gifski or should I?

I'll open a PR later today, or tomorrow. 👍

sindresorhus commented:
I noticed a tiny issue with the cancel button. If I click and hold it, drag the mouse outside the button, and then release, it will stay filled (active state), and the next time I hover it, it will still be filled.

![link style](https://user-images.githubusercontent.com/170270/51890868-e9d89b00-23cf-11e9-9e00-7d8d2712d310.gif)

boyvanamstel commented:
Yea, that happens with the show in Finder button in Gifski as well. We'll need to fix that in `CustomButton`.
